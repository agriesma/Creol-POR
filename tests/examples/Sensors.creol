
/* Abstract the network and everything behind it. */
interface Network
begin
  with Radio
    op send (in data: Data)
end

interface Radio
begin
  with Network
    op send (in data: Data)
    op send (in update: Update)
  with Controller
    op write (in data: Data)
    op read (out data: Data)
    op setPower (in power: Int)
    op setState (in state: Int)
    op getChannelStatus (out status: Int)
    op getError (out error: Int)
end

interface Controller
begin
  with Sensor
    op write (in value: Data)
end

interface Sensor
begin
  with Controller
    op setResolution (in res: Data)
    op setEncodung (in encoding: Data)
    op switchOn
    op switchOff
end

interface PassiveSensor inherits Sensor
begin
  with Controller
    op read(out value: Data)
  with Radio
    op setRate (in newRate: Duration)
end

interface ActiveSensor inherits Sensor
begin
  with Controller
    op setRate(in rate: Duration)
end

class PassiveTempSensor inherits PassiveSensor
begin
  with Controller
    op setResolution (in res: Data) == skip
    op setEncoding (in encoding: Data) == skip
    op switchOn == skip
    op switchOff == skip
    op read (out value: Data) == value := 37
end

class SimpleSensorNode(sensor: PassiveSensor, radio: Radio, size: Int,
		       initialRate: Duration)
  contracts Controller
begin
  var buffer: List[Data]
  var slots: Int
  var t0: Time
  var rate: Duration

  op init == buffer := nil; slots := size; rate := initialRate; t0 := now

  op run == read(;); await now >= t0 + rate; posit now <= t0 + rate; t0 := now;
    !run()

  op read == var temp: Data; sensor.read(; temp); posit now <= t0;
    await slots > 0; buffer := buffer |- temp; slots := slots - 1;
    if slots = 0 then packAndSend(;) end

  op packAndSend == radio.write(this -| now -| buffer;); buffer := nil;
    slots := size

  with Radio
    op setRate(in newRate: Duration) ==
      if newRate < rate && now > t0 + newRate then t0 := now; read(;) end;
      rate := newRate; posit now < t0 + rate
end

class TwoSimpleSensorNode(sensor0: PassiveSensor, sensor1: PassiveSensor,
		          radio: Radio, size: Int,
		          initialRate0: Duration, initialRate1: Duration)
  contracts Controller
begin
  var buffer: List[Data]
  var slots: Int
  var t0: Time
  var t1: Time
  var rate0: Duration
  var rate1: Duration

  op init ==
    buffer, slots, rate0, rate1, t0, t1 :=
      nil, size, initialRate0, initialRate1, now, now

  op run == !run0() ||| !run1()

  op run0 == read0(;); await now >= t0 + rate0; posit now <= t0 + rate0;
    t0 := now; !run0()

  op read0 == var temp: Data; sensor0.read(; temp); posit now <= t0;
    await slots > 0; buffer := buffer |- temp; slots := slots - 1;
    if slots = 0 then packAndSend(;) end

  op run1 == read1(;); await now >= t1 + rate1; posit now <= t1 + rate1;
    t1 := now; !run1()

  op read1 == var temp: Data; sensor1.read(; temp); posit now <= t1;
    await slots > 0; buffer := buffer |- temp; slots := slots - 1;
    if slots = 0 then packAndSend(;) end

  op packAndSend == radio.write(this -| now -| buffer;); buffer := nil;
    slots := size

  with Radio
    op setRate(in newRate: Duration) ==
      if newRate < rate0 && now > t0 + newRate then t0 := now; read0(;) end;
      rate0 := newRate; posit now < t0 + rate0
end

class ActiveSensor(controller: Controller, initialRate: Duration)
  contracts Sensor
begin
  var rate: Duration
  var t0: Time

  op init == rate, t0 := initialRate, now

  op run == await now >= t0 + rate; controller.write(37;);
    posit now <= t0 + rate; t0 := now

  with Controller
    op setRate(in newRate: Duration) ==
      if newRate < rate && now > t0 + newRate then
        controller.write(37;); t0 := now
      end;
      rate := newRate; posit now < t0 + rate
end
