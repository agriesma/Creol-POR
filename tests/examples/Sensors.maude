load creol-interpreter
mod PROGRAM is
protecting CREOL-SIMULATOR .
op init : -> Configuration [ctor] .
eq init =
< "PassiveTempSensor" : Cl | Inh: noInh, Par: noVid, Att: noSubst, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return ( emp ) > *
  < "setResolution" : Mtdname | Param: "res", Latt: "res" |-> null , "_" |-> null, Code: skip ; return ( emp ) > *
  < "setEncoding" : Mtdname | Param: "encoding", Latt: "encoding" |-> null , "_" |-> null, Code: skip ; return ( emp ) > *
  < "switchOn" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return ( emp ) > *
  < "switchOff" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return ( emp ) > *
  < "read" : Mtdname | Param: noVid, Latt: "value" |-> null , "_" |-> null, Code: "value" ::= int(37) ; return ( "value" ) > *
  < "setRate" : Mtdname | Param: "newRate", Latt: "newRate" |-> null , "_" |-> null, Code: skip ; return ( emp ) >, Ocnt: 0 >

< "SimpleSensorNode" : Cl | Inh: noInh, Par: "sensor" , "radio" , "size" , "initialRate", Att: "sensor" |-> null , "radio" |-> null , "size" |-> null , "initialRate" |-> null , "buffer" |-> null , "slots" |-> null , "t0" |-> null , "rate" |-> null, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: "buffer" ::= list(emp) ; "slots" ::= "size" ; "rate" ::= "initialRate" ; "t0" ::= now ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: "label:1" |-> null , "label:0" |-> null , "_" |-> null, Code: ( "label:0" ! "this" . "read" (emp ) )  ; ( "label:0" ? ( noVid ) )  ; ( await ">=" ( now :: "+" ( "t0" :: "rate" ) ) ) ; ( posit "<=" ( now :: "+" ( "t0" :: "rate" ) ) ) ; "t0" ::= now ; ( "label:1" ! "this" . "run" (emp ) )  ; free( "label:1" ) ; return ( emp ) > *
  < "read" : Mtdname | Param: noVid, Latt: "label:1" |-> null , "label:0" |-> null , "temp" |-> null , "_" |-> null, Code: "label:0" ! "sensor" . "read" ( emp ) ; ( "label:0" ? ( "temp" ) )  ; ( posit "<=" ( now :: "t0" ) ) ; ( await ">" ( "slots" :: int(0) ) ) ; "buffer" ::= "|-" ( "buffer" :: "temp" ) ; "slots" ::= "-" ( "slots" :: int(1) ) ; if "=" ( "slots" :: int(0) ) th ( "label:1" ! "this" . "packAndSend" (emp ) )  ; ( "label:1" ? ( noVid ) )  el skip fi ; return ( emp ) > *
  < "packAndSend" : Mtdname | Param: noVid, Latt: "label:0" |-> null , "_" |-> null, Code: "label:0" ! "radio" . "write" ( "-|" ( "this" :: "-|" ( now :: "buffer" ) ) ) ; ( "label:0" ? ( noVid ) )  ; "buffer" ::= list(emp) ; "slots" ::= "size" ; return ( emp ) > *
  < "setRate" : Mtdname | Param: "newRate", Latt: "newRate" |-> null , "label:0" |-> null , "_" |-> null, Code: if "&&" ( "<" ( "newRate" :: "rate" ) :: ">" ( now :: "+" ( "t0" :: "newRate" ) ) ) th "t0" ::= now ; ( "label:0" ! "this" . "read" (emp ) )  ; ( "label:0" ? ( noVid ) )  el skip fi ; "rate" ::= "newRate" ; ( posit "<" ( now :: "+" ( "t0" :: "rate" ) ) ) ; return ( emp ) > *
  < "write" : Mtdname | Param: "data", Latt: "data" |-> null , "_" |-> null, Code: skip ; return ( emp ) >, Ocnt: 0 >

< "TwoSimpleSensorNode" : Cl | Inh: noInh, Par: "sensor0" , "sensor1" , "radio" , "size" , "initialRate0" , "initialRate1", Att: "sensor0" |-> null , "sensor1" |-> null , "radio" |-> null , "size" |-> null , "initialRate0" |-> null , "initialRate1" |-> null , "buffer" |-> null , "slots" |-> null , "t0" |-> null , "t1" |-> null , "rate0" |-> null , "rate1" |-> null, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: "_" |-> null, Code:  ( "buffer", "slots", "rate0", "rate1", "t0", "t1" )  ::= list(emp) :: "size" :: "initialRate0" :: "initialRate1" :: now :: now ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: "label:1" |-> null , "label:0" |-> null , "_" |-> null, Code: ( "label:0" ! "this" . "run0" (emp ) )  ; free( "label:0" ) ; ( "label:1" ! "this" . "run1" (emp ) )  ; free( "label:1" ) ; return ( emp ) > *
  < "run0" : Mtdname | Param: noVid, Latt: "label:1" |-> null , "label:0" |-> null , "_" |-> null, Code: ( "label:0" ! "this" . "read0" (emp ) )  ; ( "label:0" ? ( noVid ) )  ; ( await ">=" ( now :: "+" ( "t0" :: "rate0" ) ) ) ; ( posit "<=" ( now :: "+" ( "t0" :: "rate0" ) ) ) ; "t0" ::= now ; ( "label:1" ! "this" . "run0" (emp ) )  ; free( "label:1" ) ; return ( emp ) > *
  < "read0" : Mtdname | Param: noVid, Latt: "label:1" |-> null , "label:0" |-> null , "temp" |-> null , "_" |-> null, Code: "label:0" ! "sensor0" . "read" ( emp ) ; ( "label:0" ? ( "temp" ) )  ; ( posit "<=" ( now :: "t0" ) ) ; ( await ">" ( "slots" :: int(0) ) ) ; "buffer" ::= "|-" ( "buffer" :: "temp" ) ; "slots" ::= "-" ( "slots" :: int(1) ) ; if "=" ( "slots" :: int(0) ) th ( "label:1" ! "this" . "packAndSend" (emp ) )  ; ( "label:1" ? ( noVid ) )  el skip fi ; return ( emp ) > *
  < "run1" : Mtdname | Param: noVid, Latt: "label:1" |-> null , "label:0" |-> null , "_" |-> null, Code: ( "label:0" ! "this" . "read1" (emp ) )  ; ( "label:0" ? ( noVid ) )  ; ( await ">=" ( now :: "+" ( "t1" :: "rate1" ) ) ) ; ( posit "<=" ( now :: "+" ( "t1" :: "rate1" ) ) ) ; "t1" ::= now ; ( "label:1" ! "this" . "run1" (emp ) )  ; free( "label:1" ) ; return ( emp ) > *
  < "read1" : Mtdname | Param: noVid, Latt: "label:1" |-> null , "label:0" |-> null , "temp" |-> null , "_" |-> null, Code: "label:0" ! "sensor1" . "read" ( emp ) ; ( "label:0" ? ( "temp" ) )  ; ( posit "<=" ( now :: "t1" ) ) ; ( await ">" ( "slots" :: int(0) ) ) ; "buffer" ::= "|-" ( "buffer" :: "temp" ) ; "slots" ::= "-" ( "slots" :: int(1) ) ; if "=" ( "slots" :: int(0) ) th ( "label:1" ! "this" . "packAndSend" (emp ) )  ; ( "label:1" ? ( noVid ) )  el skip fi ; return ( emp ) > *
  < "packAndSend" : Mtdname | Param: noVid, Latt: "label:0" |-> null , "_" |-> null, Code: "label:0" ! "radio" . "write" ( "-|" ( "this" :: "-|" ( now :: "buffer" ) ) ) ; ( "label:0" ? ( noVid ) )  ; "buffer" ::= list(emp) ; "slots" ::= "size" ; return ( emp ) > *
  < "setRate" : Mtdname | Param: "newRate", Latt: "newRate" |-> null , "label:0" |-> null , "_" |-> null, Code: if "&&" ( "<" ( "newRate" :: "rate0" ) :: ">" ( now :: "+" ( "t0" :: "newRate" ) ) ) th "t0" ::= now ; ( "label:0" ! "this" . "read0" (emp ) )  ; ( "label:0" ? ( noVid ) )  el skip fi ; "rate0" ::= "newRate" ; ( posit "<" ( now :: "+" ( "t0" :: "rate0" ) ) ) ; return ( emp ) > *
  < "write" : Mtdname | Param: "data", Latt: "data" |-> null , "_" |-> null, Code: skip ; return ( emp ) >, Ocnt: 0 >

< "ActiveSensor" : Cl | Inh: noInh, Par: "controller" , "initialRate", Att: "controller" |-> null , "initialRate" |-> null , "rate" |-> null , "t0" |-> null, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: "_" |-> null, Code:  ( "rate", "t0" )  ::= "initialRate" :: now ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: "label:0" |-> null , "_" |-> null, Code: ( await ">=" ( now :: "+" ( "t0" :: "rate" ) ) ) ; "label:0" ! "controller" . "write" ( int(37) ) ; ( "label:0" ? ( noVid ) )  ; ( posit "<=" ( now :: "+" ( "t0" :: "rate" ) ) ) ; "t0" ::= now ; return ( emp ) > *
  < "setRate" : Mtdname | Param: "newRate", Latt: "newRate" |-> null , "label:0" |-> null , "_" |-> null, Code: if "&&" ( "<" ( "newRate" :: "rate" ) :: ">" ( now :: "+" ( "t0" :: "newRate" ) ) ) th "label:0" ! "controller" . "write" ( int(37) ) ; ( "label:0" ? ( noVid ) )  ; "t0" ::= now el skip fi ; "rate" ::= "newRate" ; ( posit "<" ( now :: "+" ( "t0" :: "rate" ) ) ) ; return ( emp ) > *
  < "setResolution" : Mtdname | Param: "data", Latt: "data" |-> null , "_" |-> null, Code: skip ; return ( emp ) > *
  < "setEncoding" : Mtdname | Param: "data", Latt: "data" |-> null , "_" |-> null, Code: skip ; return ( emp ) > *
  < "switchOn" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return ( emp ) > *
  < "switchOff" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return ( emp ) >, Ocnt: 0 >

.
endm
