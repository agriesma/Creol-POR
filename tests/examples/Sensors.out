==========================================
reduce in PROGRAM : init .
result Configuration: 
< "ActiveSensor" : Cl | Inh: noInh, Par: "controller", "initialRate", Att:
    "controller" |-> null, "initialRate" |-> null, "rate" |-> null, "t0" |->
    null, Mtds:
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: ("rate", "t0") ::=
    "initialRate" # now ; return(emp) > *
  < "run" : Mtdname | Param: noVid, Latt: "label:0" |-> null, Code: await ">="(
    now # "+"("t0" # "rate")) ; "label:0" ! "controller" . "write"(int(37)) ;
    "label:0" ?(noVid) ; posit "<="(now # "+"("t0" # "rate")) ; "t0" ::= now ;
    return(emp) > *
  < "setEncoding" : Mtdname | Param: "data", Latt: "data" |-> null, Code: skip
    ; return(emp) > *
  < "setRate" : Mtdname | Param: "newRate", Latt: "label:0" |-> null, "newRate"
    |-> null, Code: if "&&"("<"("newRate" # "rate") # ">"(now # "+"("t0" #
    "newRate"))) th "label:0" ! "controller" . "write"(int(37)) ; "label:0" ?(
    noVid) ; "t0" ::= now el skip fi ; "rate" ::= "newRate" ; posit "<"(now #
    "+"("t0" # "rate")) ; return(emp) > *
  < "setResolution" : Mtdname | Param: "data", Latt: "data" |-> null, Code:
    skip ; return(emp) > *
  < "switchOff" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(
    emp) > *
  < "switchOn" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(
    emp) >,
  Ocnt: 0 >

< "PassiveTempSensor" : Cl | Inh: noInh, Par: noVid, Att: noSubst, Mtds:
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >
    *
  < "read" : Mtdname | Param: noVid, Latt: "value" |-> null, Code: "value" ::=
    int(37) ; return("value") > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) > *
  < "setEncoding" : Mtdname | Param: "encoding", Latt: "encoding" |-> null,
    Code: skip ; return(emp) > *
  < "setRate" : Mtdname | Param: "newRate", Latt: "newRate" |-> null, Code:
    skip ; return(emp) > *
  < "setResolution" : Mtdname | Param: "res", Latt: "res" |-> null, Code: skip
    ; return(emp) > *
  < "switchOff" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(
    emp) > *
  < "switchOn" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(
    emp) >,
  Ocnt: 0 >

< "SimpleSensorNode" : Cl | Inh: noInh, Par: "sensor", "radio", "size",
    "initialRate", Att: "buffer" |-> null, "initialRate" |-> null, "radio" |->
    null, "rate" |-> null, "sensor" |-> null, "size" |-> null, "slots" |->
    null, "t0" |-> null, Mtds:
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: "buffer" ::= list(
    emp) ; "slots" ::= "size" ; "rate" ::= "initialRate" ; "t0" ::= now ;
    return(emp) > *
  < "packAndSend" : Mtdname | Param: noVid, Latt: "label:0" |-> null, Code:
    "label:0" ! "radio" . "write"("-|"("this" # "-|"(now # "buffer"))) ;
    "label:0" ?(noVid) ; "buffer" ::= list(emp) ; "slots" ::= "size" ; return(
    emp) > *
  < "read" : Mtdname | Param: noVid, Latt: "label:0" |-> null, "label:1" |->
    null, "temp" |-> null, Code: "label:0" ! "sensor" . "read"(emp) ; "label:0"
    ?("temp") ; posit "<="(now # "t0") ; await ">"("slots" # int(0)) ; "buffer"
    ::= "|-"("buffer" # "temp") ; "slots" ::= "-"("slots" # int(1)) ; if "="(
    "slots" # int(0)) th "label:1" ! "this" . "packAndSend"(emp) ; "label:1" ?(
    noVid) el skip fi ; return(emp) > *
  < "run" : Mtdname | Param: noVid, Latt: "label:0" |-> null, "label:1" |->
    null, Code: "label:0" ! "this" . "read"(emp) ; "label:0" ?(noVid) ; await
    ">="(now # "+"("t0" # "rate")) ; posit "<="(now # "+"("t0" # "rate")) ;
    "t0" ::= now ; "label:1" ! "this" . "run"(emp) ; free("label:1") ; return(
    emp) > *
  < "setRate" : Mtdname | Param: "newRate", Latt: "label:0" |-> null, "newRate"
    |-> null, Code: if "&&"("<"("newRate" # "rate") # ">"(now # "+"("t0" #
    "newRate"))) th "t0" ::= now ; "label:0" ! "this" . "read"(emp) ; "label:0"
    ?(noVid) el skip fi ; "rate" ::= "newRate" ; posit "<"(now # "+"("t0" #
    "rate")) ; return(emp) > *
  < "write" : Mtdname | Param: "data", Latt: "data" |-> null, Code: skip ;
    return(emp) >,
  Ocnt: 0 >

< "TwoSimpleSensorNode" : Cl | Inh: noInh, Par: "sensor0", "sensor1", "radio",
    "size", "initialRate0", "initialRate1", Att: "buffer" |-> null,
    "initialRate0" |-> null, "initialRate1" |-> null, "radio" |-> null, "rate0"
    |-> null, "rate1" |-> null, "sensor0" |-> null, "sensor1" |-> null, "size"
    |-> null, "slots" |-> null, "t0" |-> null, "t1" |-> null, Mtds:
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: ("buffer", "slots",
    "rate0", "rate1", "t0", "t1") ::= list(emp) # "size" # "initialRate0" #
    "initialRate1" # now # now ; return(emp) > *
  < "packAndSend" : Mtdname | Param: noVid, Latt: "label:0" |-> null, Code:
    "label:0" ! "radio" . "write"("-|"("this" # "-|"(now # "buffer"))) ;
    "label:0" ?(noVid) ; "buffer" ::= list(emp) ; "slots" ::= "size" ; return(
    emp) > *
  < "read0" : Mtdname | Param: noVid, Latt: "label:0" |-> null, "label:1" |->
    null, "temp" |-> null, Code: "label:0" ! "sensor0" . "read"(emp) ;
    "label:0" ?("temp") ; posit "<="(now # "t0") ; await ">"("slots" # int(0))
    ; "buffer" ::= "|-"("buffer" # "temp") ; "slots" ::= "-"("slots" # int(1))
    ; if "="("slots" # int(0)) th "label:1" ! "this" . "packAndSend"(emp) ;
    "label:1" ?(noVid) el skip fi ; return(emp) > *
  < "read1" : Mtdname | Param: noVid, Latt: "label:0" |-> null, "label:1" |->
    null, "temp" |-> null, Code: "label:0" ! "sensor1" . "read"(emp) ;
    "label:0" ?("temp") ; posit "<="(now # "t1") ; await ">"("slots" # int(0))
    ; "buffer" ::= "|-"("buffer" # "temp") ; "slots" ::= "-"("slots" # int(1))
    ; if "="("slots" # int(0)) th "label:1" ! "this" . "packAndSend"(emp) ;
    "label:1" ?(noVid) el skip fi ; return(emp) > *
  < "run" : Mtdname | Param: noVid, Latt: "label:0" |-> null, "label:1" |->
    null, Code: ("label:0" ! "this" . "run0"(emp) ; free("label:0") |||
    "label:1" ! "this" . "run1"(emp) ; free("label:1")) ; return(emp) > *
  < "run0" : Mtdname | Param: noVid, Latt: "label:0" |-> null, "label:1" |->
    null, Code: "label:0" ! "this" . "read0"(emp) ; "label:0" ?(noVid) ; await
    ">="(now # "+"("t0" # "rate0")) ; posit "<="(now # "+"("t0" # "rate0")) ;
    "t0" ::= now ; "label:1" ! "this" . "run0"(emp) ; free("label:1") ; return(
    emp) > *
  < "run1" : Mtdname | Param: noVid, Latt: "label:0" |-> null, "label:1" |->
    null, Code: "label:0" ! "this" . "read1"(emp) ; "label:0" ?(noVid) ; await
    ">="(now # "+"("t1" # "rate1")) ; posit "<="(now # "+"("t1" # "rate1")) ;
    "t1" ::= now ; "label:1" ! "this" . "run1"(emp) ; free("label:1") ; return(
    emp) > *
  < "setRate" : Mtdname | Param: "newRate", Latt: "label:0" |-> null, "newRate"
    |-> null, Code: if "&&"("<"("newRate" # "rate0") # ">"(now # "+"("t0" #
    "newRate"))) th "t0" ::= now ; "label:0" ! "this" . "read0"(emp) ;
    "label:0" ?(noVid) el skip fi ; "rate0" ::= "newRate" ; posit "<"(now #
    "+"("t0" # "rate0")) ; return(emp) > *
  < "write" : Mtdname | Param: "data", Latt: "data" |-> null, Code: skip ;
    return(emp) >,
  Ocnt: 0 >
Bye.
